### PR TITLE
8267307: Introduce new client property for XAWT: xawt.mwm_decor_title

### PR DESCRIPTION
**Pull request by mkartashev**
This commit introduces a new client property xawt.mwm_decor_title implementing JDK-8267307. The property can be set prior to showing a window or after the window has been displayed, in which case the window will have to be hidden/shown (re-mapped) for the property to take effect.

The general idea is to provide control over the "decorations" part of the X11 window property called _MOTIF_WM_HINTS. Those "decoration" bits are set to 1 (XWindowAttributesData.AWT_DECOR_ALL) to show all the decorations or 0 (XWindowAttributesData.AWT_DECOR_NONE) to ask the window manager (WM) not to decorate with anything, even borders or resize handles. With xawt.mwm_decor_title property set to "true", this commit adds the ability to set the bits to 2 (XWindowAttributesData.AWT_DECOR_BORDER), which some WMs take as "decorate with only a border", thus effectively removing the window's title bar, but still leaving the resize capability.

This feature was tested and works correctly on "vanilla" Ubuntu 20.04 with the "GNOME Shell" window manager. It was also tested with Xfwm4 and KDE, where it did not have any effect; these two WMs have limited respected for the "decorations" bits of the _MOTIF_WM_HINTS window property.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8267307](https://bugs.openjdk.java.net/browse/JDK-8267307): Introduce new client property for XAWT: xawt.mwm_decor_title

### Reviewers
 * [Alexander Zvegintsev](https://openjdk.java.net/census#azvegint) (@azvegint - **Reviewer**) ⚠️ Review applies to da9fbfd800f4bc8fc26ac7fa3909a64991d07a80
 * [Sergey Bylokhov](https://openjdk.java.net/census#serb) (@mrserb - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4113/head:pull/4113` \
`$ git checkout pull/4113`

Update a local copy of the PR: \
`$ git checkout pull/4113` \
`$ git pull https://git.openjdk.java.net/jdk pull/4113/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4113`

View PR using the GUI difftool: \
`$ git pr show -t 4113`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4113.diff">https://git.openjdk.java.net/jdk/pull/4113.diff</a>

</details>

### CODE REVIEW

**Comment by mrserb** [comment]
Hello, thank you for your contribution, One initial question: what the name "mwm_" of this property actually means?

File: test/jdk/java/awt/Window/WindowTitleVisibleTest/WindowTitleVisibleTestLinuxGnome.java [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+                }
+            }
+        }
+
+        return true;
+    }
+
+    private void captureTitleBarNotVisible() {
+        runSwing( () -> {
+            titleBarImageNotVisible = robot.createScreenCapture(titleBarBounds);
```
Comment (azvegint) [inline test/jdk/java/awt/Window/WindowTitleVisibleTest/WindowTitleVisibleTestLinuxGnome.java:111]: `createScreenCapture` does not require to be called on EDT.

**Comment by azvegint** [comment]
This fix most likely will require a CSR

**Comment by mkartashev** [comment]
> This fix most likely will require a CSR

There already is one linked from the enhancement: [JDK-8267308](https://bugs.openjdk.java.net/browse/JDK-8267308)

**Comment by azvegint** [comment]
/csr

> There already is one linked from the enhancement: [JDK-8267308](https://bugs.openjdk.java.net/browse/JDK-8267308)

Sorry, I missed that.

**Review by azvegint — APPROVED** [review]

**Comment by mrserb** [comment]
Please check that the test code added in the JDK-8265005 can be adapted to verify this fix as well.

**Comment by mkartashev** [comment]
> Please check that the test code added in the JDK-8265005 can be adapted to verify this fix as well.

@mrserb Please clarify: do you suggest to have one common test code for this feature and JDK-8265005? This is certainly possible, but will contaminate the Mac test with Linux-specific window manager checking stuff, making it harder to debug failures, I think.

**Comment by mrserb** [comment]
You can reuse that test or create the new one, I just pointed out how the "same" feature was tested on mac.

File: src/java.desktop/unix/classes/sun/awt/X11/XFramePeer.java [hunk 2]
```diff
@@ -80,6 +76,38 @@ void preInit(XCreateWindowParams params) {
                      Integer.valueOf(winAttr.decorations), Boolean.valueOf(winAttr.initialResizability),
                      Boolean.valueOf(!winAttr.nativeDecor), Integer.valueOf(winAttr.initialState));
         }
+
+        registerWindowDecorationChangeListener();
+    }
+
+    private void registerWindowDecorationChangeListener() {
+        if (target instanceof javax.swing.RootPaneContainer) {
```
Comment (mrserb) [inline src/java.desktop/unix/classes/sun/awt/X11/XFramePeer.java:84]: Probably this should be done via runtime check? Overwise the swing classes will be always loaded even if swing is not used. see the usage of SunToolkit.isInstanceOf in java.awt.WIndow.java:3965

Comment (mkartashev) [reply src/java.desktop/unix/classes/sun/awt/X11/XFramePeer.java]: Thanks! Updated the code; please, have a look.

File: test/jdk/java/awt/Window/WindowTitleVisibleTest/WindowTitleVisibleTestLinuxGnome.java (hunk 1 above)
Comment (mkartashev) [reply test/jdk/java/awt/Window/WindowTitleVisibleTest/WindowTitleVisibleTestLinuxGnome.java]: Thanks! Moved `createScreenCapture()` out of EDT. Please, have a look at the updated code.

**Review by azvegint — APPROVED** [review]

File: src/java.desktop/unix/classes/sun/awt/X11/XDecoratedPeer.java [hunk 3]
```diff
@@ -1305,4 +1318,24 @@ public void handleWindowFocusOut(Window oppositeWindow, long serial) {
         }
         super.handleWindowFocusOut(oppositeWindow, serial);
     }
+
+    public static final String MWM_DECOR_TITLE_PROPERTY_NAME = "xawt.mwm_decor_title";
+
+    public final Optional<Boolean> getMWMDecorTitleProperty() {
+        Optional<Boolean> res = Optional.empty();
+
+        if (target instanceof javax.swing.RootPaneContainer) {
```
Comment (mrserb) [inline src/java.desktop/unix/classes/sun/awt/X11/XDecoratedPeer.java:1327]: Looks like this instanceof will load the Swing as well.

Comment (mkartashev) [reply src/java.desktop/unix/classes/sun/awt/X11/XDecoratedPeer.java]: Sorry, missed that one the first time. Corrected now.

**Comment by mrserb** [comment]
Looks fine, I'll run the tests.

**Pull request closed without merging**